drivers: flash: stm32: expose stm32 option byte apis

fabiobaltieri opened the pull request:
The STM32 flash driver includes some APIs to lock/unlock/set the option bytes. Make these public so that an application can use them to tweak option bytes as needed, add a read api as well.

Allows doing stuff like

```c
        uint32_t optb = 0;
        ret = flash_ex_op(DEVICE_DT_GET(DT_NODELABEL(flash)), FLASH_STM32_EX_OP_OPTB_READ, 0, &optb);
        printk("%d %08x\n", ret, optb);
```

fabiobaltieri committed:
drivers: flash: stm32: add include guards

Add missing include guards to flash_stm32.h.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

erwango reviewed (commented):
Thanks @fabiobaltieri.

Looks fine, but however, I wonder if, for consistency sake, these APIs shouldn't be made available through already existing `flash_ex_op()` api, which is there to expose flash controller specific features ?

Agreed this would be slightly more complex to implement and likely consume more footprint, but I think consistency is key for users.

Alternatively, if there are good reasons not to use `flash_ex_op()`, let's make it explicit.

fabiobaltieri commented:
> Looks fine, but however, I wonder if, for consistency sake, these APIs shouldn't be made available through already existing `flash_ex_op()` api, which is there to expose flash controller specific features ?

Oh that makes load of sense, yeah thanks for the pointer I'll move this back to draft and rework it.

fabiobaltieri committed:
drivers: flash: stm32: add STM32 option bytes extended ops

Add two new flash extended operations for reading and writing STM32
option bytes from the application code.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
@erwango I've reworked it to have two ex op codes but I don't really like the resulting code if I can be totally honest. What do you think?

gautierg-st commented on drivers/flash/Kconfig.stm32:
```suggestion
	  Enables flash extended operations that can be used to write and read
	  STM32 option bytes.
```

fabiobaltieri replied:
see, if my github account got compromised you'll be able to tell immediately as soon as you see a PR of me without at least one typo

gautierg-st reviewed (changes requested):
Small typo, otherwise LGTM

fabiobaltieri commented:
ah, needs a rebase too
